crates/polars-core/src/frame/asof_join/asof.rs

```diff
@@ -182,16 +182,102 @@ pub(super) fn join_asof_backward<T: PartialOrd + Copy + Debug>(
     out
 }
 
+pub(super) fn join_asof_nearest_with_tolerance<
+    T: PartialOrd + Copy + Debug + Sub<Output = T> + Add<Output = T> + Bounded,
+>(
+    left: &[T],
+    right: &[T],
+    tolerance: T,
+) -> Vec<Option<IdxSize>> {
+    let n_left = left.len();
+    let mut out = Vec::with_capacity(n_left);
+
+    if left.is_empty() {
+        return out;
+    }
+    if right.is_empty() {
+        out.extend(std::iter::repeat(None).take(n_left));
+        return out;
+    }
+
+    // if we know the first/last values we can leave early in many cases
+    let n_right = right.len();
+    let first_left = left[0];
+    let last_left = left[n_left - 1];
+    let r_lower_bound = right[0] - tolerance;
+    let r_upper_bound = right[n_right - 1] + tolerance;
+
+    // fast exit: guaranteed no matches
+    if (r_lower_bound > last_left) || (r_upper_bound < first_left) {
+        out.extend(std::iter::repeat(None).take(n_left));
+        return out;
+    }
+
+    for &val_l in left {
+        // no match detection
+        if val_l < r_lower_bound {
+            // less than all values on right, move to next immediately
+            out.push(None);
+            continue;
+        } else if val_l > r_upper_bound {
+            // greater than all values on right, nothing else match; return early
+            out.extend(std::iter::repeat(None).take(n_left - out.len()));
+            return out;
+        }
+
+        // we might have a match
+        let mut offset: IdxSize = 0;
+        let mut dist = tolerance;
+        let mut found_window = false;
+        for &val_r in right {
+            // haven't reached the window, keep looking
+            if val_l > val_r + tolerance {
```

Comment:
Can you assign a temp_val for `val_r + tolerance`? It is used more often.

Reply:
We have a `val_l + tolerance` farther down, which I think you may have mistook for `val_r + tolerance`--this is only used once. However, `val_l + tolerance` doens't change inside the inner loop so we can temp that one.